src/Translator.php

```diff
@@ -0,0 +1,582 @@
+        foreach ($this->modulePaths as $modulePath) {
+            foreach (glob($this->getYmlLangDirectory($modulePath) . '/*.yml') as $path) {
+                $rawYaml = file_get_contents($path);
+                $this->originalYaml[$path] = Yaml::parse($rawYaml);
+            }
+        }
+        $this->log('Finished backing up ' . count($this->originalYaml) . ' yaml files');
+    }
+
+    /**
+     * Set mtime to a year ago so that transifex will see these as obsolete
+     */
+    private function setJsonAndYmlFileTimes()
+    {
+        $date = date('YmdHi.s', strtotime('-1 year'));
+        foreach ($this->modulePaths as $modulePath) {
+            $name = $this->getModuleName($modulePath);
+            $this->log("Setting file mtime to a past date for <info>{$name}</info>");
+            $ymlLang = $this->getYmlLangDirectory($modulePath);
+            if ($ymlLang) {
+                $this->exec("find $ymlLang -type f \( -name \"*.yml\" \) -exec touch -t $date {} \;");
+            }
+            foreach ($this->getJSLangDirectories($modulePath) as $jsLangDir) {
+                $this->exec("find $jsLangDir -type f \( -name \"*.json*\" \) -exec touch -t $date {} \;");
+            }
+        }
+    }
+
+    /**
+     * Update sources from transifex
+     */
+    private function transifexPullSource()
+    {
+        foreach ($this->modulePaths as $modulePath) {
+            // ensure .tx/config is up to date
+            $contents = file_get_contents("$modulePath/.tx/config");
+            if (strpos($contents, '[o:') === false) {
+                $this->exec('tx migrate', $modulePath);
+                // delete .bak files created as part of tx migrate
+                foreach (scandir("$modulePath/.tx") as $filename) {
+                    if (pathinfo($filename, PATHINFO_EXTENSION) !== 'bak') {
+                        continue;
+                    }
+                    $this->log("Deleting $modulePath/.tx/$filename");
+                    unlink("$modulePath/.tx/$filename");
+                }
+            }
+            // pull from transifex
+            $this->exec("tx pull -a -s -t -f --minimum-perc={$this->txMinimumPerc}", $modulePath);
+        }
+    }
+
+    /**
+     * Merge any missing keys from old yaml content into yaml files
+     */
+    private function mergeYaml(): void
+    {
+        $this->log('Merging local yaml files');
+        foreach ($this->originalYaml as $path => $contentYaml) {
+            // If there are any keys in the original yaml that are missing now, add them back in.
+            if (!file_exists($path)) {
+                $this->log("$path does not exist, continuing");
+                continue;
+            }
+            $rawYaml = file_get_contents($path);
+            $parsedYaml = Yaml::parse($rawYaml);
+            $contentYaml = $this->arrayMergeRecursive($contentYaml, $parsedYaml);
+            // Write back to local
+            file_put_contents($path, Yaml::dump($contentYaml));
```

Comment:
```suggestion
            if (file_exists($path)) {
                $rawYaml = file_get_contents($path);
                $parsedYaml = Yaml::parse($rawYaml);
                $contentYaml = $this->arrayMergeRecursive($contentYaml, $parsedYaml);
            }
            // Write back to local
            file_put_contents($path, Yaml::dump($contentYaml));
```
This is how it was in cow, and seems more appropriate. If transifex deleted the file we don't want to commit that change because it will mean we're deleting those translations! Instead, restore the file from the backup.

Same with the json please.

Reply:
Updated